### PR TITLE
Move the standard library to a separate workspace 

### PR DESCRIPTION
**Pull request by bjorn3**
This ensures that the Cargo.lock packaged for it in the rust-src component is up-to-date, allowing rust-analyzer to run cargo metadata on the standard library even when the rust-src component is stored in a read-only location as is necessary for loading crates.io dependencies of the standard library.

This also simplifies tidy's license check for runtime dependencies as it can now look at all entries in library/Cargo.lock without having to filter for just the dependencies of runtime crates. In addition this allows removing an exception in check_runtime_license_exceptions that was necessary due to the compiler enabling a feature on the object crate which pulls in a dependency not allowed for the standard library.

While cargo workspaces normally enable dependencies of multiple targets to be reused, for the standard library we do not want this reusing to prevent conflicts between dependencies of the sysroot and of tools that are built using this sysroot. For this reason we already use an unstable cargo feature to ensure that any dependencies which would otherwise be shared get a different -Cmetadata argument as well as using separate build dirs.

This doesn't change the situation around vendoring. We already have several cargo workspaces that need to be vendored. Adding another one doesn't change much.

There are also no cargo profiles that are shared between the root workspace and the library workspace anyway, so it doesn't add any extra work when changing cargo profiles.

### CODE REVIEW

**Comment by bjorn3** [comment]
Previous attempt at #76533. Most of the arguments against it back then are no longer true as we already have to deal with many workspaces anyway.

**Comment by bjorn3** [comment]
cc @Veykril as this should allow `rust-analyzer.cargo.sysrootQueryMetadata` to become true by default in future rust-analyzer versions.

**Comment by bjorn3** [comment]
This should also enable `-Zbuild-std` to lock the versions of stdlib dependencies to the same versions as used when building rustc rather than generating a lockfile on the fly with the latest version and discarding it again after the build.

**Comment by Veykril** [comment]
> cc @Veykril as this should allow `rust-analyzer.cargo.sysrootQueryMetadata` to become true by default in future rust-analyzer versions.

Looking at the changes this is still not enough I think. We need the standard library to ship a `Cargo.toml` ***and*** a corresponding lockfile, the latter being what metadata creates right now which prevents the setting from being enabled by default. (the former is semi solved by us invoking metadata on the dummy `sysroot` crate as an entry point)

**Comment by bjorn3** [comment]
Yes, the rust-src component now contains `library/Cargo.lock`. I've also checked that running `cargo metadata` in the library dir works.

**Comment by workingjubilee** [comment]
I should also mention that the _failure_ to divide between workspaces has made things much more confusing for many people trying to understand how the standard library and/or compiler are compiled and which dependencies go where. And understanding what, exactly, goes into the standard library, and thus is linked into programs, is somewhat relevant for understanding [legal obligations](https://www.apache.org/licenses/LICENSE-2.0).

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me

File: src/bootstrap/src/core/metadata.rs [hunk 1]
```diff
@@ -87,5 +87,8 @@ fn workspace_members(build: &Build) -> Vec<Package> {
     };
 
     // Collects `metadata.packages` from all workspaces.
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply src/bootstrap/src/core/metadata.rs]: All usages of `in_tree_crates`, `crate_or_deps` and `crates` seem to only need the root and library workspace. I've updated the comment.

**Comment by glandium** [comment]
Does it mean -Zbuild-std doesn't need its virtual workspace anymore?

**Comment by workingjubilee** [comment]
?
https://github.com/rust-lang/rust/pull/128534#issuecomment-2265135337

Try implementing removing that?

**Comment by bjorn3** [comment]
Not familiar with the implementation of -Zbuild-std, but likely.

**Comment by glandium** [comment]
I'll check, because I have downstream fore-ports of https://github.com/rust-lang/cargo/pull/8834 and https://github.com/rust-lang/rust/pull/78790 that I would be happy to get rid of.

**Comment by tgross35** [comment]
Cc relevant zulip discussion https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Project.20goal.3A.20.60-Zbuild-std.60.3F

**Comment by weihanglo** [comment]
> Does it mean -Zbuild-std doesn't need its virtual workspace anymore?

https://github.com/rust-lang/cargo/pull/14358

**Pull request merged**